### ISSUE TITLE
TS2339: Property does not exist on type 'never'

### ISSUE THREAD

**Issue by AHHusam** [microsoft/TypeScript#55350]
### 🔎 Search Terms

TS2339

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about it

### 💻 Code

```ts
interface Foo {
  Foo: {
    f: string
  }
}

interface Bar {
  Bar: {
    b: string
  }
}

interface Baz {
  Baz: {
    z: string
  }
}

type All = Foo | Bar | Baz;

const message: All = {
  Bar: {
    b: 'wow'
  }
};

if ('Foo' in message) {
  console.log(message.Foo);
} else if ('Bar' in message) {
  console.log(message.Bar.b)
} else if ('Baz' in message) {
  console.log(message.Baz) // Property 'Baz' does not exist on type 'never'. ts(2339)
}

```

### 🙁 Actual behavior

TS complains that 'Baz' doesn't exist on type 'never'.

If I re-order the if-else blocks, always the last one will complain that it doesn't exist on type 'never'.

### 🙂 Expected behavior

TS shouldn't complain because I'm defining the allowed types.

**Comment by MartinJohns** [comment]
You forgot to fill out the issue template.

This is working as intended. Assignment narrows the type of the variable, so the variable is not typed `All` but `Bar`. So after you checked that the `Bar` property does not exist, the type checker is sure that it can't be any other type, ending up `never`. And checking if a property exists on `never` makes no sense.

You can work around the narrowing of the variable by using a type assertion: `{ ... } as All`.
